zeus/device/cpu/__init__.py

```diff
@@ -6,11 +6,34 @@
 
 from __future__ import annotations
 
+from typing import Literal
+import os
+
 from zeus.device.cpu.common import CPUs, ZeusCPUInitError
 from zeus.device.cpu.rapl import rapl_is_available, RAPLCPUs
 
 _cpus: CPUs | None = None
 
+def get_current_cpu_index(pid: int | Literal["current"] = "current") -> int:
+    """Retrieves the specific CPU index (socket) where the given PID is running.
+
+    If no PID is given or pid is "current", the CPU index returned is of the CPU running the current process.
+
+    Note: Linux schedulers can preempt and reschedule processes to different CPUs. To prevent this from happening
+    during monitoring, use `taskset` to pin processes to specific CPUs.
+    """
+    if pid == "current":
+        pid = os.getpid()
+        
+    cpu_core = None
+    with open(f"/proc/{pid}/stat", "r") as stat_file:
+        cpu_core = int(stat_file.read().split()[38])
```

Comment:
I think it is guaranteed to exist.

```
Description:	CPU topology files that describe a logical CPU's relationship
		to other cores and threads in the same physical package.

		One cpuX directory is created per logical CPU in the system,
		e.g. /sys/devices/system/cpu/cpu42/.
```

[sys/devices/system/cpu/smt/control](url)

Also this page seems to say that the 39th field is the CPU core number.

```
(39) CPU number last executed on.

rt_priority %u (since Linux 2.5.19; was %lu before Linux 2.6.22)
```

[https://linux.die.net/man/5/proc](url)

Reply:
Sounds good! Looks like a pretty strong guarantee.